alexandria/core/serializers.py

```diff
@@ -10,30 +10,47 @@
 class BaseSerializer(serializers.ModelSerializer):
     created_at = serializers.DateTimeField(read_only=True)
 
+    def is_valid(self, *args, **kwargs):
+        user = self.context["request"].user
+        # Prime data so the validators are called (and default values filled
+        # if client didn't pass them.)
+        self.initial_data.setdefault("created_by_user", user.username)
+        self.initial_data.setdefault("created_by_group", self._default_group())
+        self.initial_data.setdefault("modified_by_user", user.username)
+        self.initial_data.setdefault("modified_by_group", self._default_group())
+        return super().is_valid(*args, **kwargs)
+
     def validate(self, *args, **kwargs):
         validated_data = super().validate(*args, **kwargs)
 
         self.Meta.model.check_permissions(self.context["request"])
         if self.instance is not None:
             self.instance.check_object_permissions(self.context["request"])
 
-        user = self.context["request"].user
-        default_group = user.group if not isinstance(user, AnonymousUser) else None
-
-        validated_data["modified_by_user"] = user.username
-        if not validated_data.get("modified_by_group"):
-            validated_data["modified_by_group"] = default_group
-        if self.instance is None:
-            validated_data["created_by_user"] = user.username
-            if not validated_data.get("created_by_group"):
-                validated_data["created_by_group"] = default_group
         return validated_data
 
+    def validate_created_by_user(self, value):
+        # Created by user is set once on creation and cannot be modified after
+        user = self.context["request"].user
+        return self.instance.created_by_user if self.instance else user.username
+
     def validate_created_by_group(self, value):
-        return self._validate_group(value, "created_by_group")
+        # Created by group can be set on creation, then must remain constant
+        if self.instance:
+            # can't change created_by_group on existing instances
+            return self.instance.created_by_group
+        else:
+            return self._validate_group(value, "created_by_group")
+
+    def validate_modified_by_user(self, value):
```

Comment:
What purpose is there that this can be sent by the client?

Reply:
None really. This bit is just refactoring from the previous "big" `validate()` method. But thinking about it again, I suppose this could be a lot simpler - fix coming in a minute.

Do you think we could set those fields as `read_only` in `Meta`?

Reply:
> Do you think we could set those fields as read_only in Meta?

I was thinking something like that. But I'm not sure if they then get dropped before saving.

Reply:
Yep that's my understanding as well. Only works for created_at, as it's handled by the model itself